### PR TITLE
FEFaceEvaluation: Select general path for exterior cells with constraints

### PR DESCRIPTION
**Pull request by bergbauer**
In ECL face loops with constraints `read_dof_values()` should only select the contiguous path if possible.

### CODE REVIEW

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -3377,11 +3377,36 @@ FEEvaluationBase<dim, n_components_, Number, is_face, VectorizedArrayType>::
...
-            [is_face ? this->dof_access_index :
-                       internal::MatrixFreeFunctions::DoFInfo::dof_access_cell]
-            [this->cell] >= internal::MatrixFreeFunctions::DoFInfo::
-                              IndexStorageVariants::contiguous)
+              [is_face ?
+                 this->dof_access_index :
+                 internal::MatrixFreeFunctions::DoFInfo::dof_access_cell]
+              [this->cell] >= internal::MatrixFreeFunctions::DoFInfo::
+                                IndexStorageVariants::contiguous &&
+          exterior_cells_are_contiguous)
```
[the remark this answers is not in the document]
Comment (bergbauer) [reply include/deal.II/matrix_free/fe_evaluation.h]: I have uploaded two versions (one with a separate if statement for `is_face`). Which one do you prefer? @kronbichler

[the remark this answers is not in the document]
Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h]: I like the two versions much better. I would prefer version 1, as the separate `is_face` statement does not give a lot of improvement. But I would be happy with the second versions as well, if others prefer that.

[the remark this answers is not in the document]
Comment (bergbauer) [reply include/deal.II/matrix_free/fe_evaluation.h]: Let's go with version 1 then.

**Review by peterrum — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Pull request merged**